token-client/src/main/java/com/sap/cloud/security/client/DefaultHttpClientFactory.java

```diff
@@ -24,27 +34,65 @@
 public class DefaultHttpClientFactory implements HttpClientFactory {
 
 	private static final Logger LOGGER = LoggerFactory.getLogger(DefaultHttpClientFactory.class);
+	// reuse ssl connections
+	final ConcurrentHashMap<String, SslConnection> sslConnectionPool = new ConcurrentHashMap<>();
+	final List<String> httpClientsCreated = new ArrayList<>();
+	static final int MAX_CONNECTIONS_PER_ROUTE = 4; // 2 is default
+	static final int MAX_CONNECTIONS = 20;
+	private static final int DEFAULT_TIMEOUT = (int) TimeUnit.SECONDS.toMillis(5);
+	private final RequestConfig timeoutConfig;
+
+	public DefaultHttpClientFactory() {
+		timeoutConfig = RequestConfig.custom()
+				.setConnectTimeout(DEFAULT_TIMEOUT)
+				.setConnectionRequestTimeout(DEFAULT_TIMEOUT)
+				.setSocketTimeout(DEFAULT_TIMEOUT)
+				.build();
+	}
 
 	@Override
 	public CloseableHttpClient createClient(ClientIdentity clientIdentity) throws HttpClientException {
-		LOGGER.warn("In productive environment provide well configured HttpClientFactory service");
-		if (clientIdentity != null && clientIdentity.isCertificateBased()) {
-			LOGGER.debug("Setting up HTTPS client with: certificate: {}\n", clientIdentity.getCertificate());
-			SSLContext sslContext;
+		if (clientIdentity != null) {
+			if (httpClientsCreated.contains(clientIdentity.getId())) {
+				LOGGER.warn("Application has already created HttpClient for {}, please check.", clientIdentity.getId());
+			}
+			httpClientsCreated.add(clientIdentity.getId());
```

Comment:
This code is not thread safe. A simple solution would be to make `httpClientsCreated` a HashSet and synchronize it with `Collections.synchronizedSet(...)`.

However, the overall flow does not capture the case when `clientIdentity == null` but still HttpClients get created.

Reply:
done 👍